src/node_locks.cc

```diff
@@ -0,0 +1,738 @@
+
+  Local<Promise::Resolver> waiting_promise;
+  if (!Promise::Resolver::New(context).ToLocal(&waiting_promise)) {
+    return;
+  }
+  Local<Promise::Resolver> released_promise;
+  if (!Promise::Resolver::New(context).ToLocal(&released_promise)) {
+    return;
+  }
+
+  args.GetReturnValue().Set(released_promise->GetPromise());
+
+  LockManager* manager = GetCurrent();
+  {
+    Mutex::ScopedLock scoped_lock(manager->mutex_);
+
+    // Register cleanup hook for the environment only once
+    if (manager->registered_envs_.insert(env).second) {
+      env->AddCleanupHook(LockManager::OnEnvironmentCleanup, env);
+    }
+
+    auto lock_request = std::make_unique<LockRequest>(env,
+                                                      waiting_promise,
+                                                      released_promise,
+                                                      callback,
+                                                      resource_name,
+                                                      lock_mode,
+                                                      client_id,
+                                                      steal,
+                                                      if_available);
+    // Steal requests get priority by going to front of queue
+    if (steal) {
+      manager->pending_queue_.emplace_front(std::move(lock_request));
+    } else {
+      manager->pending_queue_.push_back(std::move(lock_request));
+    }
+  }
+
+  manager->ProcessQueue(env);
+}
+
+void LockManager::Query(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Isolate* isolate = env->isolate();
+  HandleScope scope(isolate);
+  Local<Context> context = env->context();
+
+  Local<Promise::Resolver> resolver;
+  if (!Promise::Resolver::New(context).ToLocal(&resolver)) {
+    return;
+  }
+  args.GetReturnValue().Set(resolver->GetPromise());
+
+  Local<Object> result = Object::New(isolate);
+  Local<Array> held_list = Array::New(isolate);
+  Local<Array> pending_list = Array::New(isolate);
+  LockManager* manager = GetCurrent();
+
+  {
+    Mutex::ScopedLock scoped_lock(manager->mutex_);
+
+    uint32_t index = 0;
+    for (const auto& resource_entry : manager->held_locks_) {
+      for (const auto& held_lock : resource_entry.second) {
+        if (held_lock->env() == env) {
+          Local<Object> lock_info =
+              CreateLockInfoObject(isolate,
+                                   context,
+                                   held_lock->name(),
+                                   held_lock->mode(),
+                                   held_lock->client_id());
+          if (lock_info.IsEmpty()) {
+            return;
+          }
+          held_list->Set(context, index++, lock_info).Check();
+        }
+      }
+    }
+
+    index = 0;
+    for (const auto& pending_request : manager->pending_queue_) {
+      if (pending_request->env() == env) {
+        Local<Object> lock_info =
```

Comment:
Same here... if the `Local<Object> lock_info;` is moved outside of the for loops then it can just be reused here.